Setup Go builds and tests in the CI

torives opened the pull request:
This PR adds new steps to build and test Go code and a new workflow to format and lint it. I've added a single `GraphQLService` test to ensure the CI is running tests. If you feel there's an opportunity to write more, let me know and I'll gladly do it.

Example runs:
- [Format error](https://github.com/cartesi/rollups-node/actions/runs/6357098174/job/17267779583?pr=107)
- [Linter error](https://github.com/cartesi/rollups-node/actions/runs/6385162227/job/17329318970?pr=107). Below is an image demonstrating the annotation created when a linter error is found:
![image](https://github.com/cartesi/rollups-node/assets/2274568/9673f912-2201-4270-a0c8-7aead817633e)


Closes #108

torives committed:
ci: add Go-related workflows and steps

torives committed:
feat: add go binary to Dockerfile

torives committed:
refactor: rename binary to cartesi-rollups-node

gligneul commented on .github/workflows/build.yml:
```diff
@@ -78,6 +78,15 @@ jobs:
+
+      - name: Run Go tests
+        working-directory: ${{ github.workspace }}
+        run: go test ./internal/pkg/services
```
```suggestion
        run: go test ./...
```

To run all tests

gligneul commented on internal/pkg/services/graphql-service_test.go:
```diff
@@ -0,0 +1,49 @@
+package services
```
I noticed that the GraphQL file is using `-` in its filename. By convention, we should use only `_`.

torives replied:
I'll shorten the names to avoid having to use a separator at all and leave the `build` and `workflows` directories as is

gligneul commented on internal/pkg/services/graphql-service_test.go:
```diff
@@ -0,0 +1,49 @@
+func setupEnvVars() {
+	abs, _ := filepath.Abs("../../../offchain/target/debug")
+	os.Setenv("PATH", abs)
+	os.Setenv("POSTGRES_PASSWORD", "pw")
```
We removed this variable in https://github.com/cartesi/rollups-node/commit/a3c95f3ed69518464d9d695966c9fd14b7937edf

gligneul replied:
We can remove it; the graphQL will try to connect to a local Postgres, but it will take some time to fail because of the backoff.

gligneul commented on internal/pkg/services/graphql-service_test.go:
```diff
@@ -0,0 +1,49 @@
+		exit := make(chan error)
+
+		go func() {
+			<-time.After(100 * time.Millisecond)
```
I think we don't need this sleep.

gligneul reviewed (commented):
Looks great! Just minor comments.

torives commented on internal/pkg/services/graphql-service_test.go:
```diff
@@ -0,0 +1,49 @@
+package services
```
I'll come up with something to check this automatically. I constantly forget to add the copyright 🙄

endersonmaia commented on build/Dockerfile:
```diff
@@ -67,6 +71,15 @@ RUN cargo chef cook --release --recipe-path recipe.json
+ARG GO_BUILD_PATH
+WORKDIR ${GO_BUILD_PATH}
+COPY . .
+RUN go install -ldflags "-s -w" ./cmd/rollups-node
```
Shouldn't the binary share the same prefix as others?

`cartesil-rollups-node`

torives commented on .github/workflows/build.yml:
```diff
@@ -78,6 +78,15 @@ jobs:
           path: offchain/schema.graphql
           if-no-files-found: error
 
+      - name: Install Go
```
I left it here to avoid rebuilding the Rust binaries. It's not the prettiest thing but it's efficient. In the future, after we remove the Rust parts the workflow will become a lot simpler.

endersonmaia commented on .github/workflows/go-code-quality.yml:
```diff
@@ -0,0 +1,34 @@
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v3
```
A lot of actions were updated to use latest javascript versions for its runtime.

```suggestion
      - uses: actions/checkout@v4
```

This update could be done on another PR, updating all actions to latest version, for ex.

torives replied:
I'll address that in a new PR. Thanks for the heads up 😉

torives commented:
I've simplified the `assess-go-code-quality` job as I found out the linter had a `gofmt` option, so I removed the extra step. Also added a few more linters:

`exhaustive`: check exhaustiveness of enum switch statements
`goconst`: Finds repeated strings that could be replaced by a constant
`godox`: Tool for detection of FIXME, TODO and other comment keywords
`gomnd`: An analyzer to detect magic numbers.
`lll`: Reports long lines
`misspell`: Finds commonly misspelled English words in comments

marcelstanley commented on internal/pkg/services/graphql-service.go:
```diff
@@ -30,7 +30,10 @@ func (g GraphQLService) Start(ctx context.Context) error {
 		<-ctx.Done()
 		fmt.Printf("%v: %v\n", g.String(), ctx.Err())
-		cmd.Process.Signal(syscall.SIGTERM)
+		if err := cmd.Process.Signal(syscall.SIGTERM); err != nil {
+			msg := "%v: failed to send SIGTERM to %v\n"
+			fmt.Printf(msg, g.String(), binaryName)
+		}
```
Is this related to the CI workflow or something else?

torives replied:
Fixes a linter error for not handling the `error` returned from the function

marcelstanley replied:
Perhaps a separate commit would also fit better here

torives commented on .github/workflows/rust-code-quality.yml:
```diff
@@ -4,7 +4,7 @@ name: Assess Rust code quality
   push:
     paths:
-      - '.github/workflows/code-quality.yml'
+      - '.github/workflows/rust-code-quality.yml'
```
Agreed 😉

torives committed:
ci: fix workflow file path

The file was renamed in commit fb059d but the workflow wasn't updated accordingly

torives commented:
I've added new changelog entries for the changes in this PR. I promise this is the last review request 😅

gligneul commented on CHANGELOG.md:
```diff
@@ -7,10 +7,19 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+### Added
+
+- Added `cartesi-rollups-node` Go binary as a single entrypoint to execute all Cartesi Node services
+- Added Go-related workflows to the CI
```
I would remove the CI-related changes from the CHANGELOG because they are irrelevant to external parties. Us developers can use `git log` to track these changes.

torives committed:
fix: handle the return of all functions

This problem was signaled by the linter tool introduced in 01c62e
